File: src/main/java/com/werfad/utils/EditorUtils.java
package com.werfad.utils;

import com.intellij.openapi.editor.Editor;
import com.intellij.openapi.editor.LogicalPosition;
import com.intellij.openapi.editor.ScrollingModel;
import com.intellij.openapi.editor.VisualPosition;
import com.intellij.openapi.util.TextRange;
import org.jetbrains.annotations.NotNull;

import java.awt.*;

public final class EditorUtils {
    @NotNull
    public static TextRange getVisibleRangeOffset(@NotNull Editor editor) {
        ScrollingModel scrollingModel = editor.getScrollingModel();
        Rectangle visibleArea = scrollingModel.getVisibleArea();

        LogicalPosition startLog = editor.xyToLogicalPosition(new Point(0, visibleArea.y));
        LogicalPosition lastLog = editor.xyToLogicalPosition(new Point(0, visibleArea.y + visibleArea.height));

        int startOff = editor.logicalPositionToOffset(startLog);
        int endOff = editor.logicalPositionToOffset(new LogicalPosition(lastLog.line + 1, lastLog.column));
        return new TextRange(startOff, endOff);
    }

    @NotNull
    public static Point offsetToXYCompat(@NotNull Editor editor, int offset) {
        return offsetToXYCompat(editor, offset, false, false);
    }

    @NotNull
    public static Point offsetToXYCompat(@NotNull Editor editor, int offset, boolean leanForward, boolean beforeSoftWrap) {
        VisualPosition visualPosition = editor.offsetToVisualPosition(offset, leanForward, beforeSoftWrap);
        return editor.visualPositionToXY(visualPosition);
    }
}

File: src/main/kotlin/com/werfad/utils/EditorUtils.kt
package com.werfad.utils

import com.intellij.openapi.editor.Editor
import com.intellij.openapi.editor.LogicalPosition
import com.intellij.openapi.util.TextRange
import java.awt.Point

fun Editor.getVisibleRangeOffset(): TextRange {
    val scrollingModel = scrollingModel
    val visibleArea = scrollingModel.visibleArea
    val startLog = xyToLogicalPosition(Point(0, visibleArea.y))
    val lastLog = xyToLogicalPosition(Point(0, visibleArea.y + visibleArea.height))
    val startOff = logicalPositionToOffset(startLog)
    val endOff = logicalPositionToOffset(LogicalPosition(lastLog.line + 1, lastLog.column))
    return TextRange(startOff, endOff)
}

fun Editor.offsetToXYCompat(
    offset: Int,
    leanForward: Boolean = false,
    beforeSoftWrap: Boolean = false
): Point {
    val visualPosition = offsetToVisualPosition(offset, leanForward, beforeSoftWrap)
    return visualPositionToXY(visualPosition)
}